store/src/java/com/zimbra/cs/service/mail/SendMsg.java

```diff
@@ -133,9 +137,27 @@ public Element handle(Element request, Map<String, Object> context, QName respQn
                String identityId = msgElem.getAttribute(MailConstants.A_IDENTITY_ID, null);
                String dataSourceId = msgElem.getAttribute(MailConstants.A_DATASOURCE_ID, null);
                String draftId = msgElem.getAttribute(MailConstants.A_DRAFT_ID, null);
-               ItemId iidDraft = draftId == null ? null : new ItemId(draftId, zsc);
                boolean sendFromDraft = msgElem.getAttributeBool(MailConstants.A_SEND_FROM_DRAFT, false);
 
+               // get the identity by id and check if the message is send by using persona for other account
+               // As the draft is always part of  auth user's context, append the authuser accountId with draftId
+               // so that the draft will be deleted when message sent successfully in case of persona.
+               // because in case of sendAs and sendOnBehalfOf the draftId pattern is accountId:itemId.
+               if (identityId != null) {
+                   Identity identity = Provisioning.getInstance().get(zsc.getAuthToken().getAccount(), Key.IdentityBy.id, identityId);
+                   if (identity != null && zsc.getAuthToken().getAccount() != null && !zsc.getAuthToken().isAdmin()
+                           && !zsc.getAuthToken().getAccount().getBooleanAttr(Provisioning.A_zimbraAllowAnyFromAddress, false)
+                           && !StringUtil.isNullOrEmpty(identity.getAttr(Provisioning.A_zimbraPrefFromAddressType))
+                           && (identity.getAttr(Provisioning.A_zimbraPrefFromAddressType).equals(RightConsts.RT_sendAs) ||
+                                   identity.getAttr(Provisioning.A_zimbraPrefFromAddressType).equals(RightConsts.RT_sendOnBehalfOf))
+                           && !StringUtil.isNullOrEmpty(identity.getAttr(Provisioning.A_zimbraPrefFromAddress))
+                           && !zsc.getAuthToken().getAccount().getMail().equals(identity.getAttr(Provisioning.A_zimbraPrefFromAddress))) {
+                       if(draftId != null && draftId.indexOf(ItemIdentifier.ACCOUNT_DELIMITER) == -1) {
+                           draftId = zsc.getAuthToken().getAccount().getId() + ItemIdentifier.ACCOUNT_DELIMITER + draftId;
+                       }
+                   }
+               }
+               ItemId iidDraft = draftId == null ? null : new ItemId(draftId, zsc);
```

Comment:
This if is unreadable, here is a sample simplified version, 

 AuthToken authToken = zsc.getAuthToken();
               Account authAcct = getAuthenticatedAccount(zsc);
               
               if (identityId != null) {
                   Identity identity = Provisioning.getInstance().get(authAcct, Key.IdentityBy.id, identityId);
                   if (identity != null)  {
                       String fromAddrPrefType =  identity.getAccount().getPrefFromAddressTypeAsString();
                       String fromAddrPrefValue   = identity.getAccount().getPrefFromAddress();
                       if (!StringUtil.isNullOrEmpty(fromAddrPrefType) && !zsc.getAuthToken().isAdmin() 
                              && !authAcct.isAllowAnyFromAddress() && !StringUtil.isNullOrEmpty(fromAddrPrefValue)
                              && !authAcct.getMail().equals(fromAddrPrefValue)) {
                           
                           if(draftId != null && draftId.indexOf(ItemIdentifier.ACCOUNT_DELIMITER) == -1) {
                               draftId = authAcct.getId() + ItemIdentifier.ACCOUNT_DELIMITER + draftId;
                           }
                       }
                   }
               }

I would recommend replacing even the above with a method with a descriptive name that tells what it is doing.